python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -402,13 +423,29 @@ def is_reconciled(
             if asset_graph.is_source(parent.asset_key):
                 continue
 
-            if (
-                self.get_latest_materialization_record(
+            # For performance, we try to only call get_latest_materialization on unpartitioned
+            # assets. To do so, we reverse the order of operations based on the partitioning of
+            # the current asset.
+            if asset_partition.partition_key is None:
+                latest_materialization_record = self.get_latest_materialization_record(
```

Comment:
Would it make sense to push this logic down into `materialization_exists`? It seems like this logic isn't specific to notion of whether an asset is reconciled?

Reply:
mm so more like:

```
def materialization_exists_after_parent(
    child: AssetKeyPartitionKey,
    parent: Optional[AssetKeyPartitionKey],
)
```
?

not unreasonable, although the current `materialization_exists` function is conceivably useful in other contexts whereas the `materialization_exists_after_parent` is less generally-applicable. I think I'll just factor it out and have both of these functions exist (as you're right this is some heavy logic to have in the is_reconciled function)